### PR TITLE
plot,plotter: properly handle Log-Y scale in Histogram

### PR DESCRIPTION
**Pull request by sbinet**
Fixes gonum/plot#377.

Please take a look.

### CODE REVIEW

**Comment by eaburns** [comment]
The golden image has only one labeled tick mark on the y-axis, so it's not possible to tell the scale. Shouldn't there be at least two, or is that a separate but?

**Comment by eaburns** [comment]
Sorry for butting in again with little context, but it also looks like the h1 values at -2 don't show in the golden plot either. What happened to them?

**Comment by sbinet** [comment]
>  but it also looks like the h1 values at -2 don't show

ah! quite right...

**Comment by sbinet** [comment]
the values at -2 are at the bottom.
`Histogram` should probably implement either `GlyphBoxer` or "artificially" pull down the `ymin` value `DataRange` returns... arf!

**Comment by sbinet** [comment]
PTAL.

I ended up modifying `Histogram.DataRange`, keeping track of both `ymin` and `ylow`, where `ymin` is the smallest Y value and `ylow` the smallest non-zero Y value.
if `ymin` is zero, `DataRange` returns `0.5*ylow` as `ymin`, to make sure the smallest bin is still displayed.

open to a more robust algorithm, though...

File: plotter/histogram.go [hunk 1]
```diff
@@ -106,6 +115,16 @@ func (h *Histogram) DataRange() (xmin, xmax, ymin, ymax float64) {
...
 		}
+		if bin.Weight < ymin {
+			ymin = bin.Weight
+		}
+		if 0 < bin.Weight && bin.Weight < ylow {
+			ylow = bin.Weight
+		}
+	}
+	if ymin == 0 {
+		// reserve a bit of space for the smallest bin to be displayed still.
```
Comment (kortschak) [inline plotter/histogram.go]: s/reserve/Reserve/

File: plotter/histogram.go [hunk 2]
```diff
@@ -106,6 +115,16 @@ func (h *Histogram) DataRange() (xmin, xmax, ymin, ymax float64) {
 		if bin.Weight > ymax {
 			ymax = bin.Weight
 		}
+		if bin.Weight < ymin {
+			ymin = bin.Weight
+		}
+		if 0 < bin.Weight && bin.Weight < ylow {
```
Comment (kortschak) [inline plotter/histogram.go]: This reads as a range check; bin.Weight in (0, ylow). But I don't think this is what you are doing; you are checking whether bin.Weight is non-zero and only using it then. Is this correct? If so, please rephrase this as `if bin.Weight != 0 && bin.Weight < ylow {`. If it's a check that bin.Weight is positive non-zero, then replace the `!=` in that with a `>`.

**Comment by sbinet** [comment]
@decibelcooper 
> Also, what about adding a flag to Histogram for whether or not to suppress zeros? Perhaps it's better to have this off by default.

yeah... this has bothered me as well.
perhaps we should bite the bullet and just go with introducing a thin `HistogramLogY` type that wraps `Histogram` but has the zero-suppression built-in.

Commit 9ffa89d adds a special mode to `Histogram` to explictly support Y log-scale.
Less invasive but also less automated.

**Comment by sbinet** [comment]
ping?

**Comment by kortschak** [comment]
I don't understand the rationale for the first commit. We don't import "log" in plot/axis.go (or anywhere in plot).

**Comment by sbinet** [comment]
the `func log(...)` function is quite annoying (to me at least) when I sprinkle code with `log.Printf("boo")` statements during my debugging sessions.

I can remove it and send it as part of another PR. (or not.)

File: plotter/histogram.go [hunk 3]
```diff
@@ -32,6 +32,9 @@ type Histogram struct {
 	// LineStyle is the style of the outline of each
 	// bar of the histogram.
 	draw.LineStyle
+
+	// LogY enables support for histograms displayed with a Y log-scale.
```
Comment (kortschak) [inline plotter/histogram.go]: This comment is not quite right; `// LogY allows rendering with a log-scaled Y axis.` (or something similar), but also a brief explanation of what it is doing to make this possible.

**Comment by kortschak** [comment]
>I can remove it and send it as part of another PR. (or not.)

SGTM

File: plotter/histogram.go [hunk 4]
```diff
@@ -77,25 +77,34 @@ func (h *Histogram) Plot(c draw.Canvas, p *plot.Plot) {
...
 }
 
 // DataRange returns the minimum and maximum X and Y values
 func (h *Histogram) DataRange() (xmin, xmax, ymin, ymax float64) {
-	xmin = math.Inf(1)
+	xmin = math.Inf(+1)
 	xmax = math.Inf(-1)
+	ymin = math.Inf(+1)
 	ymax = math.Inf(-1)
+	ylow := math.Inf(+1) // smallest non-zero y value
```
[the remark this answers is not in the document]
Comment (sbinet) [reply plotter/histogram.go]: done.

File: plotter/histogram.go (hunk 3 above)
Comment (sbinet) [reply plotter/histogram.go]: done.

**Comment by sbinet** [comment]
(had to force-push b/c of the `logf` commit removal...)

File: plotter/histogram.go (hunk 3 above)
Comment (kortschak) [reply plotter/histogram.go]: It would still be nice to have an explanation of what it does.

Comment (sbinet) [reply plotter/histogram.go]: tentatively: aca1101

**Comment by decibelcooper** [comment]
> > I can remove it and send it as part of another PR. (or not.)
> 
> SGTM

Too bad... I, too, thought having a function called `log` is needlessly cumbersome.  Yay for the merge, though.

**Comment by sbinet** [comment]
@decibelcooper don't worry, it's been removed from the top-level (but local) namespace of the `plot` package:
- https://github.com/gonum/plot/pull/502

:)
